### PR TITLE
 Example that shows how to convert query result into rust struct #2959 

### LINKED ISSUE #2959: Example to show how to convert query result into rust struct
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Add an example that shows how to convert query result into rust struct.

**Describe the solution you'd like**
Add an example.

**Describe alternatives you've considered**
Implementing a macro to convert record batch into rust struct. 
Performance wise, this should have less overhead when implemented properly, but it takes some time to get it right. In the mean time, I think my solution is the most straight forward way to do this.

**Additional context**
Here is my implementation.
https://github.com/thomas-k-cameron/arrow-datafusion/blob/to_struct/datafusion-examples/examples/deserialize_to_struct.rs
If this is ok, I'd love to create a pull request.

I needed to convert the query result into rust struct for my project: I initially tried to implement a macro but after a while I discovered that you can actually do this using functions that are already implemented by the library.  It took me some time to figure this out, and I think this use case is relevant to many people.
I also saw a repository that is trying to implement a way to turn record batches into rust struct.

It would be very nice if you could let me know if I'm getting this wrong.

### PR DESCRIPTION
**Pull request by thomas-k-cameron**
# Which issue does this PR close?
Closes #2959 .

 # Rationale for this change
Add an example that shows how to convert query result into rust struct.
Details/Background in the issue.

# What changes are included in this PR?
New example.

# Are there any user-facing changes?
No.

### CODE REVIEW

**Comment by thomas-k-cameron** [comment]
Fixed all the stuff mentioned by the tests

File: datafusion-examples/examples/deserialize_to_struct.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use datafusion::error::Result;
+use datafusion::prelude::*;
+use serde::Deserialize;
+
+/// This example shows that it is possible to convert query results into Rust structs .
```
Comment (alamb) [inline datafusion-examples/examples/deserialize_to_struct.rs:22]: 👍

File: datafusion-examples/examples/deserialize_to_struct.rs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            let df = ctx
+                .sql("SELECT int_col, double_col FROM alltypes_plain")
+                .await?;
+
+            df.show().await?;
+
+            df.collect().await?
+        };
+        // converts it to serde_json type and then convert that into Rust type
+        let list =
```
Comment (alamb) [inline datafusion-examples/examples/deserialize_to_struct.rs:65]: 👍  this is very cool

**Review by alamb — APPROVED** [review]
I tried it out -- and it looks great to me.

```
cargo run --example deserialize_to_struct
...

+---------+------------+
| int_col | double_col |
+---------+------------+
| 0       | 0          |
| 1       | 10.1       |
| 0       | 0          |
| 1       | 10.1       |
| 0       | 0          |
| 1       | 10.1       |
| 0       | 0          |
| 1       | 10.1       |
+---------+------------+
[
    Data {
        int_col: 0,
        double_col: 0.0,
    },
    Data {
        int_col: 1,
        double_col: 10.1,
    },
    Data {
        int_col: 0,
        double_col: 0.0,
    },
    Data {
        int_col: 1,
        double_col: 10.1,
    },
    Data {
        int_col: 0,
        double_col: 0.0,
    },
    Data {
        int_col: 1,
        double_col: 10.1,
    },
    Data {
        int_col: 0,
        double_col: 0.0,
    },
    Data {
        int_col: 1,
        double_col: 10.1,
    },
]
```
👍 

Thanks @thomas-k-cameron

File: datafusion-examples/examples/deserialize_to_struct.rs (hunk 2 above)
Comment (thomas-k-cameron) [reply datafusion-examples/examples/deserialize_to_struct.rs:65]: Thanks :D

**Review by xudong963 — APPROVED** [review]
Cool example! @thomas-k-cameron

**Pull request merged**